presto-main/src/main/java/com/facebook/presto/sql/planner/LiteralInterpreter.java

```diff
@@ -144,7 +148,17 @@ else if (value.equals(Double.POSITIVE_INFINITY)) {
         }
 
         Signature signature = FunctionRegistry.getMagicLiteralFunctionSignature(type);
-        Expression rawLiteral = toExpression(object, FunctionRegistry.type(type.getJavaType()));
+        Expression rawLiteral = null;
+
+        if (type instanceof ArrayType || type instanceof MapType || type instanceof RowType) {
```

Comment:
Instead of doing this we should handle this in 3 cases:
array: rewrite this as a FunctionCall to the array constructor function, similar to how VARBINARY is handled above. I think it would look something like this
```
block = readBlock((Slice) object);
expressions = new ArrayList();
for (int i = 0; i < block.positions(); i++) {
expressions.add(toExpression(block.get(i), type.getElementType()));
}
return new FunctionCall("array_constructor", expressions);
```

map: similar to array, but convert it to a call to the `map()` function

row: I think you can just remove this case, since currently there's no way to construct a row literal